### ISSUE TITLE
Save and load?

### ISSUE THREAD

**Issue by lachesis** [Byron/dua-cli#137]
Is there a way to save the scan result and reimport it for interactive view? This would be similar to `ncdu`'s `-o` and `-f` options.

**Comment by Byron** [comment]
That would be interesting for some use-cases for sure and quite straightforward to implement I presume if one uses standard Rust serialization.

**Comment by CouldBeThis** [comment]
Not that anyone asked me but I do occasionally use [`pdu`](https://github.com/KSXGitHub/parallel-disk-usage)'s `--json-output` and `--json-input` to save and load results. You can produce different visualizations for the same data.

My main simple use case is that I have a bunch of old spinning HDDs connected to a low resource computer. I added HDDs slowly over time as I needed space so there was no scheme of what files were where. Too chaosy. So it was required to learn how space was being used then plan a way to move things around.

- Takes a long time even in rust so in this way I can set a task to collect the data then come back and sort through it later. 
- Don't have to keep a terminal/ssh/tmux or any other kind of session active to preserve the information
- If you close a window you don't have to start completely fom scratch as I did on a few occasions.
- The json files can be moved to a better computer which is more pleasant to work on. 

It's kind of like using `locate` instead of `find`. `locate` periodically creates a database then lets you very quickly search through it, whereas `find` starts from scratch every time. If your file-set is relatively stable the speed can be an acceptable trade off. 

And since it's json I guess you can do all kinds of things with it. 

I think it would be cool if `dua` could do this because it's the one I like the best but tbh it is also acceptable to use the other available tools. As an intermediary, loading the jsons generated by `pdu` would be cool because `dua` does a much superior job of visualizing. But only once in a very long while with a very specific use case.

**Comment by Byron** [comment]
Definitely CC @piotrwach :).

**Comment by piotrwach** [comment]
I also have few use cases where having ability to load a file system walk from a file would be very useful:

1. I work with remote machines where I don't want to run `dua`. I could instead run some shell command that would list all the files that I could then load in `dua` on my machine. E.g.

```
$ ssh remote_host find /home -ls > listing.txt
$ dua i --from-file listing.txt
```

2. I sometimes need to analyse what is taking space in my Amazon S3 bucket. For that I would like to run something similar as in 1) e.g. 

```
$ s3cmd ls --recursive "s3://mybucket/*" > listing.txt
$ dua i --from-file listing.txt
```

Unfortunately the formats of listing.txt in both cases are quite different:

find:
```
$ find . -ls
19692507        0 drwxr-xr-x    7 pwach            staff                 224  9 Nov 07:39 .
19692508        8 -rw-r--r--    1 pwach            staff                1000  9 Nov 07:39 ./1000bytes
19692511        0 drwxr-xr-x    3 pwach            staff                  96  9 Nov 07:39 ./empty-dir
19692512        0 -rw-r--r--    1 pwach            staff                   0  9 Nov 07:39 ./empty-dir/.gitkeep
```

vs 

s3cmd
```
$ s3cmd ls --recursive "s3://mybucket/*"
2019-06-14 17:08         123  s3://mybucket/a/1.png
2019-06-15 17:08         456  s3://mybucket/b/2.png
```

For what it is worth, I have changed `dua` to achieve 2) already, but since it is not universal / probably a bit out of scope of this project I think we need more universal way of doing this.

**Comment by ipstone** [comment]
ncdu has the options to output/input indexed folders:
```
ncdu <options> <directory>

Options:
  -h,--help                  This help message
  -q                         Quiet mode, refresh interval 2 seconds
  -v,-V,--version            Print version
  -x                         Same filesystem
  -e                         Enable extended information
  -t NUM                     Number of threads to use
  -r                         Read only
  -o FILE                    Export scanned directory to FILE
  -f FILE                    Import scanned directory from FILE
  -0,-1,-2                   UI to use when scanning (0=none,2=full ncurses)
  --si                       Use base 10 (SI) prefixes instead of base 2
  --exclude PATTERN          Exclude files that match PATTERN
  -X, --exclude-from FILE    Exclude files that match any pattern in FILE
  -L, --follow-symlinks      Follow symbolic links (excluding directories)
  --exclude-caches           Exclude directories containing CACHEDIR.TAG
  --exclude-kernfs           Exclude Linux pseudo filesystems (procfs,sysfs,cgroup,...)
  --confirm-quit             Confirm quitting ncdu
  --color SCHEME             Set color scheme (off/dark/dark-bg)
  --ignore-config            Don't load config files

Refer to `man ncdu` for the full list of options.
```

which are quite handy for working/analyzing files on the remote server etc. - which are my use cases to manage large/complex folder/files. If dua can implement similar input/output format that will be terrific!

**Comment by voltagex** [comment]
I'd also use this, with the caveat that'd be really good if the format was simpler than ncdu's. I throw GNU find -printf output into a database to be able to search across about 150TB of files over 3 machines.

**Comment by CouldBeThis** [comment]
Another option would be [`qdirstat/doc/GG-file-format`](https://github.com/shundhammer/qdirstat/blob/master/doc/cache-file-format.txt). According to the developer, it is used by other applications.

Example from the docs:

```
[qdirstat 2.0 cache file]
# Generated by qdirstat-cache-writer
# Do not edit!
#
# Type  path                            size     uid   gid  perm.       mtime      <optional fields>

D /work/home/sh/src/qdirstat            4096    1000  1000  0775        0x65e0ce47
# Device: /dev/nvme0n1p5

F       .qmake.stash                    739     1000  1000  0664        0x65915c05
F       .gitignore                      22      1000  1000  0664        0x5fb9043d
F       LICENSE                         18092   1000  1000  0664        0x4ba94fed
F       qdirstat.pro.user               20372   1000  1000  0664        0x622b43a9
F       qdirstat.pro                    832     1000  1000  0664        0x6151907d
F       README.md                       45874   1000  1000  0664        0x65ce5618
F       Makefile                        35154   1000  1000  0664        0x65915c05
D /work/home/sh/src/qdirstat/src        20480   1000  1000  0775        0x65d7ba63
F       OpenDirDialog.cpp               9411    1000  1000  0664        0x65b3e4b5
F       ui_mime-category-config-page.h  12390   1000  1000  0664        0x65931d62
F       HeaderTweaker.h                 4988    1000  1000  0664        0x6585900b
```

I have been more using this program recently because of this feature. It isn't ideal to need the GUI but overall is more efficient because the cache can be pregenerated.

It would be extra cool to interoperate with `qdirstat` though, because I think it is the best of the GUIs. Combining the speed of `dua` for crawling the disk, and the cli interface, with the option of switching to GUI when the situation calls for it would be tres cool.

**Comment by Byron** [comment]
I really like the idea and even think there'd be nothing wrong to support multiple output formats. Maybe that's where a lot of value is, even though the first format should ideally be read & write from the get-go.
